keps/sig-api-machinery/00xx-publish-crd-openapi.md

```diff
@@ -0,0 +1,140 @@
+---
+kep-number: 0
+title: Publish CRD OpenAPI
+authors:
+  - "@roycaihw"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@apelisse"
+  - "@liggitt"
+  - "@mbohlool"
+  - "@sttts"
+approvers:
+  - "@liggitt"
+  - "@sttts"
+editor: TBD
+creation-date: 2019-02-07
+last-updated: 2019-02-11
+status: provisional
+see-also:
+  - [Validation for CustomResources design doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md)
+---
+
+# Publish CRD OpenAPI
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+* [Proposal](#proposal)
+* [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+In CustomResourceDefinition (CRD) we allow CRD author to define OpenAPI v3 schema, to
+enable server-side [validation for CustomResources (CR)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md).
+The validation schema format is compatible for creating OpenAPI documentation for CRs,
+which can be used by clients like kubectl to perform client-side validation
+(e.g. `kubectl create` and `kubectl apply`),
+schema explanation (`kubectl explain`), and client generation.
+This KEP purposes using the OpenAPI v3 schema to create and publish OpenAPI
+documentation for CRs.
+
+## Motivation
+
+Publishing CRD OpenAPI enables client-side validation, schema explanation and
+client generation for CRs. It covers the gap between CRD and native Kubernetes
+APIs, which already support OpenAPI documentation.
+
+Publishing CRD OpenAPI is also noted as potential followup in [Validation for CustomResources Implementation Plan](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md#implementation-plan).
+
+### Goals
+
+* For every CRD that we serve, publish Paths (operations that we support on
+  resource and subresources) and Definitions (for both CR object and CR list
+  object) in OpenAPI documentation to fully demonstrate the existence of the
+  API.
+* For CRDs with schema defined, the CR object Definition should
+  include both CRD schema and native Kubernetes ObjectMeta and
+  TypeMeta properties.
+* For CRDs without schema, the CRD object definition will be as
+  complete as possible while still maintaining compatibility with the openapi
+  spec and with supported kubernetes components
+
+## Proposal
+
+Ref: https://github.com/kubernetes/kubernetes/pull/71192
```

Comment:
I think the KEP should probably also describe
* what you're going to do (convert from v3 -> v2 + merge, anything else?)
* where the new code will live
* where the new code will be called from / used at run time

Reply:
added a brief design proposal

trying to save review bandwidth as the topic has been long standing since custome resource validation was added